### ISSUE TITLE
Handling sparse categorical data

### ISSUE THREAD

**Issue by evancollins1** [bayesian-optimization/BayesianOptimization#519]
I am working with Bayesian Optimization. Before using it to guide the selection process of my real-world wet-lab chemical experiments where I'll have much more flexibility, I want to evaluate its effectiveness on a prior training dataset, where I retroactively use Bayesian Optimization to assess how it would have navigated the landscape. However, this training dataset (i.e., search space) is not the canonical use-case for Bayesian Optimization for the following reasons:

1. Some of the variables are categorical; to better represent the categorical variables rather than a simple one-hot encoding, I have provided embeddings in tSNE to better capture relative differences. The model, however, still ultimately should suggest one of the provided discrete categories.

2. It is sparse. For context, each row represents a different chemical mixture, where a single functional target value is measured for each mixture (i.e., 'target_value'). I have a total of 7 different variables to learn from, i.e.,

    - component1_structure (3 tSNE dimensions, but effectively categorical)
    
    - component1_ratio (1 dimension, continuous)
    
    - component1_weight (1 dimension, continuous)
    
    - component2_structure (2 tSNE dimensions, but effectively categorical)
    
    - component2_ratio (1 dimension, continuous)
    
    - component3_ratio (1 dimension, continuous)
    
    - component4_ratio (1 dimension, continuous)

This combinatorial space is theoretically very large. However, I only have 1801 different mixtures (i.e., combinations of these variables) with functional data; I want the model to strictly explore/suggest among these 1801 mixtures where I have functional data.

3. Rather than sequential optimization, each round of experiments will have multiple (i.e., 10) different mixtures to test. Hence, as described in this [paper](https://www.nature.com/articles/s41586-021-03213-y?fromPaywallRec=false), I incorporate the Kriging believer algorithm. My goal is to demonstrate that Bayesian Optimization to search this training data is effective; hence, ultimately motivating using it for an actual real-world dataset dynamically (where the second constraint described above fortunately won't apply). Nevertheless, I want to conceive of a way to use Bayesian Optimization for this preexisting training data.

Here is an implementation example on Colab: https://colab.research.google.com/drive/19cY54d6M7jYl1fRDpL7oGVPF832JU247?usp=sharing

My current strategy as you'll see on the Colab is using a weighted distance metric to basically identify the unique point (among the 1801 not already selected) closest to what the model suggests. However, I have doubts this is a sound approach.

Suggestions on how to improve this strategy and/or code implementations would be very helpful.

Thanks in advance!

**Comment by till-m** [comment]
Hey @evancollins1,

this is a pretty complex situation. I'll just drop my 2 cents to every point you mentioned, since you didn't ask any specific question.

> Some of the variables are categorical; to better represent the categorical variables rather than a simple one-hot encoding, I have provided embeddings in tSNE to better capture relative differences. The model, however, still ultimately should suggest one of the provided discrete categories.

If I'm understanding right, you're using tSNE on OHE categorical data, individually per feature? That doesn't sound very meaningful to me, intuitively. Wouldn't you just end up with singular clusters in the tSNE space, with each cluster equivalent to one category? Especially considering you're relying on distances between those clusters for the GP, which are devoid of meaning. Or maybe I misunderstand?

> This combinatorial space is theoretically very large. However, I only have 1801 different mixtures (i.e., combinations of these variables) with functional data; I want the model to strictly explore/suggest among these 1801 mixtures where I have functional data.

I think how large this space is depends a lot on the cardinality of the categorical data. Just considering the continuous data, it doesn't sound too bad considering you have 1800 data points.

> Rather than sequential optimization, each round of experiments will have multiple (i.e., 10) different mixtures to test. Hence, as described in this [paper](https://www.nature.com/articles/s41586-021-03213-y?fromPaywallRec=false), I incorporate the Kriging believer algorithm. My goal is to demonstrate that Bayesian Optimization to search this training data is effective; hence, ultimately motivating using it for an actual real-world dataset dynamically (where the second constraint described above fortunately won't apply). Nevertheless, I want to conceive of a way to use Bayesian Optimization for this preexisting training data.

I looked at the asynchronous parallel optimization problem a bit when implementing the acqusition functions, and I found [lying](https://bayesian-optimization.github.io/BayesianOptimization/master/reference/acquisition/ConstantLiar.html#bayes_opt.acquisition.ConstantLiar) to be a much better strategy in practice than believing, so much so that I ended up dropping the believer from the PR despite having fully implemented it. You could also try Thompson Sampling, which is implemented in [the notebook](https://bayesian-optimization.github.io/BayesianOptimization/master/acquisition_functions.html) (though it is quite computationally expensive).

> My current strategy as you'll see on the Colab is using a weighted distance metric to basically identify the unique point (among the 1801 not already selected) closest to what the model suggests. However, I have doubts this is a sound approach.

I'm curious how this weighted distance metric was derived. If you have a good idea on how to construct a meaningful distance function over your space, you could try using it as a kernel for the GP regressor. Similarly, if you don't have a good idea, it might be better to use the fitted GPs kernel as a distance metric.

I hope this is helpful.

**Comment by evancollins1** [comment]
Thank you @till-m for your answers (and patience with the post, as I understand it raised many questions). I am new to this field.

As for tSNE, more broadly, I thought that encoding categorical data in some representative latent space may be more informative than simple one-hot encoding, namely because the pairwise distances in latent space capture degree of similarity between the discrete categories. Is this reasoning valid? I was motivated by [this paper](https://www.nature.com/articles/s41598-020-60652-9) but my implementation/reasoning may be invalid. I believe this [package](https://github.com/b-shields/edbo/blob/master/examples/deoxyfluorination_optimization/optimization.ipynb) also encodes categorical data (e.g., chemical structures) in latent space, though didn't know how to implement this in this package. 

As for the 1801 datapoint (i.e., chemical mixtures) search space I mentioned, I guess my question is if this `bayes_opt` package allows for the search space to be constrained?, i.e., select combinations of variables are the only options that the model can consider, meaning that the allowable search space is sparse. Practically speaking, I noticed how `bayes_opt` expects `pbounds` for each variable, such as below...

`optimizer = BayesianOptimization(f=objective_function, pbounds=pbounds, verbose=2, random_state=987234)`

... but what if not every combination of variables can be considered (i.e., only the 1801 specific combinations can be considered when selecting points). 

Lastly, with respect to the weighted distance metric, how could I use it as a kernel for the GP regressor?

Appreciate your help!

**Comment by till-m** [comment]
To check, you want to do two things:
1. Evaluate BO as an approach to finding promising candidates by testing its performance on existing data
2. Apply BO in a practical setting, if (1) works well: Look for candidates and experimentally test them. Since you test batches at a time, you want the optimizer to give a set of candidates.

Is that correct?

> As for tSNE, more broadly, I thought that encoding categorical data in some representative latent space may be more informative than simple one-hot encoding, namely because the pairwise distances in latent space capture degree of similarity between the discrete categories.

There is a chance I am missing something. Here's what I think you're doing: You have a categorical feature (e.g. `fruit`) with values (e.g. `apple`, `banana` and `orange`). You OH-encode it into `1, 0, 0`, `0, 1, 0`, `0, 0, 1` vectors respectively. The tSNE will then just assign all `apple`s to one point, all `banana`s to a different point etc. There will not be a meaningful distance between them, because tSNE has no idea how similar `apple`s are to `banana`s. It only knows that an `apple` is the same as another `apple` and that two `apple`s differ the same amount from two `banana`s. It does not know whether an `apple` is more different to a `banana` than it is to an `orange`. This is more or less the same information that's present in the OHE (depending on the exact distance metric that you use to calculate the distance of the OHE vectors, if it's unweighted, then an `apple` is as different from a `banana` as it is from an `orange`). In short, without additional information this is not meaningful.

> [this paper](https://www.nature.com/articles/s41598-020-60652-9)
> this [package](https://github.com/b-shields/edbo/blob/master/examples/deoxyfluorination_optimization/optimization.ipynb)

I searched the paper for "tsne" and didn't find anything, if you point me to where they discuss the encoding I can have another look. I also didn't find exactly where the package talks about encoding.

> As for the 1801 datapoint (i.e., chemical mixtures) search space I mentioned, I guess my question is if this bayes_opt package allows for the search space to be constrained?

If I understand correctly what you want, the way to go here is to evaluate the acquisition function on all viable points (i.e. the 1801 datapoints) and select the highest-value one. That's both much faster, easier, etc. than anything else. You will still need to set pbounds, since we did not foresee this case and I don't think the optimizer can be constructed without them. For what it's worth, this approach seems scientifically sound to me.

> Lastly, with respect to the weighted distance metric, how could I use it as a kernel for the GP regressor? - is there example code somewhere that I could learn from?

See [here](https://scikit-learn.org/stable/modules/gaussian_process.html#kernels-for-gaussian-processes) for the sklearn docs on GPR kernels. How does the function look exactly?
Once you have defined your `kernel` object, you can call `optimizer.set_gp_params(kernel=kernel)` to use it.

If you have categorical features, the best way to deal with them AFAIK is to make the kernel aware of it, see [here](https://arxiv.org/abs/1805.03463). The bad news is, this is not implemented in this library. You can sort-of monkeypatch it, but it's tricky. If you feel comfortable in python and want to do this, you can check out this [comment](https://github.com/bayesian-optimization/BayesianOptimization/issues/191#issuecomment-629828879).

**Comment by evancollins1** [comment]
Hi @till-m,

Thanks again for your thoughts! 

Yes - you are correct that those are my two objectives.

For tSNE, to clarify, my tSNE embeddings were determined by analyzing 1024 data features (i.e., additional information) of the discrete categories. To continue the fruit analogy, I have feature data for 1024 different characteristics of the fruits, such that each discrete fruit category can be mapped in this 1024-dimensional space. I then use tSNE to reduce into 3 dimensions, with the idea being that pairwise distance in this space captures extent of similarity between two fruits.

As for my first objective of analyzing the preexisting dataset...

> If I understand correctly what you want, the way to go here is to evaluate the acquisition function on all viable points (i.e. the 1801 datapoints) and select the highest-value one. That's both much faster, easier, etc. than anything else. You will still need to set pbounds, since we did not foresee this case and I don't think the optimizer can be constructed without them. For what it's worth, this approach seems scientifically sound to me.

...that's good to know that this approach is theoretically faster and easier. How would you recommend I set pbounds and enforce the model must select only from the discrete 1801 points? I understand the concept that the acquisition function needs to evaluate the 1801 points and select the highest one; do you have any code suggestions on implementing this in this package?

> If you have categorical features, the best way to deal with them AFAIK is to make the kernel aware of it, see [here](https://arxiv.org/abs/1805.03463). The bad news is, this is not implemented in this library. You can sort-of monkeypatch it, but it's tricky. If you feel comfortable in python and want to do this, you can check out this https://github.com/bayesian-optimization/BayesianOptimization/issues/191#issuecomment-629828879.

Lastly, thank you for pointing me to the resources for encoding categorical features. Is this package able to handle categorical variables like this example: temperature values where the acceptable set of temperatures is [10, 15, 20, 25, 30], i.e,. 10 to 30 degrees by intervals by 5.

Thanks in advance!

**Comment by till-m** [comment]
If the above does not do the trick and you think you might need more help with the data science part of the problem, feel free to let me know. It sounds like an interesting problem and wouldn't mind having a look at it to see if there's the possibility of some sort of collaboration, depending on the scope and direction of the project (I assume this is academic in nature?)

**Comment by evancollins1** [comment]
Thanks again for your response @till-m,

> Moreover, if the optimizer suggests you sample a point that doesn't match neatly into one of the two categories -- which is likely to happen -- can you do that with your experiment?

This is a good point about decoding back to categorical space that I also considered. I'd be curious to hear your opinion about a sister package to this one called `edbo` that I've mentioned in my prior comments. In this [example notebook](https://github.com/b-shields/edbo/blob/master/examples/deoxyfluorination_optimization/optimization.ipynb) for `ebdo`, they encode categorical variables using more descriptors. After reading the paper and code, however, I still was not sure how they decode back into categorical space (also raised in this [issue here](https://github.com/b-shields/edbo/issues/23)). If you have any quick thoughts on this that'd be great.

> It sounds like an interesting problem and wouldn't mind having a look at it to see if there's the possibility of some sort of collaboration, depending on the scope and direction of the project

Great! Yes, this is academic. I'd be happy to chat more about this and discuss a potential collaboration. Feel free to send me an email at evanc@mit.edu if you are interested and we can schedule a time to chat; thanks, looking forward to it

**Comment by till-m** [comment]
Hey @evancollins1,

> This is a good point about decoding back to categorical space that I also considered. I'd be curious to hear your opinion about a sister package to this one called edbo that I've mentioned in my prior comments. In this [example notebook](https://github.com/b-shields/edbo/blob/master/examples/deoxyfluorination_optimization/optimization.ipynb) for ebdo, they encode categorical variables using more descriptors. After reading the paper and code, however, I still was not sure how they decode back into categorical space (also raised in this https://github.com/b-shields/edbo/issues/23). If you have any quick thoughts on this that'd be great.

I'm not sure what `descriptor` is but if I am reading the EDBO paper right, they're using OHE for the categorical variables. With that it should be straightforward to map back to categories. But maybe I misread?

> Feel free to send me an email at [evanc@mit.edu](mailto:evanc@mit.edu) if you are interested and we can schedule a time to chat; thanks, looking forward to it

Excellent, I'll shoot you an email later :)
